backend/src/v5/routes/sso/aad/aad.js

```diff
@@ -0,0 +1,165 @@
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+const { checkAadConfig, getUserDetailsAndValidateEmail } = require('../../../middleware/dataConverter/inputs/sso/aad');
+const { Router } = require('express');
+const Users = require('../../../processors/users');
+const { getAuthenticationCodeUrl } = require('../../../utils/sso/aad');
+const { respond } = require('../../../utils/responder');
+const { templates } = require('../../../utils/responseCodes');
+const { validateSignUpData } = require('../../../middleware/dataConverter/inputs/users');
+
+const authenticateRedirectUri = 'http://localhost/api/v5/sso/aad/authenticate-post';
+const signupRedirectUri = 'http://localhost/api/v5/sso/aad/signup-post';
+// Frontend uri for signup form
+const sigupUri = '';
+
+const authenticate = (req, res) => {
+	try {
+		const params = { redirectUri: authenticateRedirectUri, state: sigupUri };
+		const authenticationCodeUrl = getAuthenticationCodeUrl(params);
+		res.redirect(authenticationCodeUrl);
+	} catch (err) {
+		/* istanbul ignore next */
+		respond(req, res, err);
+	}
+};
+
+const authenticatePost = (req, res) => {
+	try {
+		res.redirect(req.query.state);
+	} catch (err) {
+		/* istanbul ignore next */
+		respond(req, res, err);
+	}
+};
+
+const signup = (req, res) => {
+	try {
+		const { body } = req;
+		const params = {
+			redirectUri: signupRedirectUri,
+			state: JSON.stringify({
+				username: body.username,
+				countryCode: body.countryCode,
+				company: body.company,
+				mailListAgreed: body.mailListAgreed,
+			}),
+		};
+
+		const authenticationCodeUrl = getAuthenticationCodeUrl(params);
+		res.redirect(authenticationCodeUrl);
+	} catch (err) {
+		/* istanbul ignore next */
+		respond(req, res, err);
+	}
+};
+
+const signupPost = async (req, res) => {
+	try {
+		await Users.signUp(req.body);
+		respond(req, res, templates.ok);
+	} catch (err) {
+		/* istanbul ignore next */
+		respond(req, res, err);
+	}
+};
+
+const establishRoutes = () => {
+	const router = Router({ mergeParams: true });
+
+	/**
+	* @openapi
+	* /sso/aad/authenticate:
+	*   get:
+	*     description: Redirects the user to Microsoft's authentication page and calls authenticate-post endpoint upon successful authentication
+	*     tags: [Aad]
+	*     operationId: authenticate
+	*/
+	router.get('/authenticate', checkAadConfig, authenticate);
+
+	/**
+	 * @openapi
+	 * /sso/aad/authenticate-post:
+	 *   get:
+	 *     description: Redirects the user to 3D Repo's SSO signup page
+	 *     tags: [Aad]
+	 *     operationId: authenticatePost
+	 */
+	router.get('/authenticate-post', checkAadConfig, authenticatePost);
```

Comment:
i think `authenticate-post` needs be a constant so you don't risk renaming this and not renaming the redirect URI